File: (2) not_oop/main.cpp
#include <iostream>
#include <memory>
#include <vector>

class Pet {
public:
    // With `virtual`, a vtable gets inseted into the class, increasing its size
    // and tightly couplinfg the class with the idea of inheritance.
    virtual ~Pet() = default;
    virtual void Speak() const noexcept = 0;
};


class Cat : public Pet {
public:
    void Speak() const noexcept override {
        std::cout << "Meow" << std::endl;
    }
};


class Dog : public Pet {
public:
    void Speak() const noexcept override {
        std::cout << "Woof" << std::endl;
    }
};


void speak_from_func(Pet& pet) {
    // This does 3 dereferences:
    // 1. Pointer to the class.
    // 2. Pointer to the vtable.
    // 3. Pointer to the function.
    pet.Speak();
}


std::vector<std::unique_ptr<Pet>> create_pets() {
    std::vector<std::unique_ptr<Pet>> pets;
    pets.reserve(2);
    // These have to be allocated.
    pets.push_back(std::make_unique<Dog>());
    pets.push_back(std::make_unique<Cat>());
    return pets;
}


int main() {
    // Normal construction.
    auto dog = Dog();
    dog.Speak();
    auto cat = Cat();
    cat.Speak();

    // Passing to functions.
    speak_from_func(dog);
    speak_from_func(cat);

    // Iterating over pets.
    auto pets = create_pets();
    for(const auto& pet : pets) {
        // This is 2 dereferences.
        // 1. To vtable.
        // 2. To function.
        pet->Speak();
    }
}

File: (2) not_oop/src/main.rs
#![allow(dead_code)]

use enum_dispatch::enum_dispatch;

#[enum_dispatch]
trait Voice {
    fn speak(&self);
}

// Dog.

struct Dog;

impl Dog {
    pub fn new() -> Self {
        Dog {}
    }
}

impl Voice for Dog {
    fn speak(&self) {
        println!("Woof");
    }
}

// Cat.

struct Cat;

impl Cat {
    pub fn new() -> Self {
        Cat {}
    }
}

impl Voice for Cat {
    fn speak(&self) {
        println!("Meow");
    }
}

#[enum_dispatch(Voice)]
enum Pet {
    Dog(Dog),
    Cat(Cat),
}

fn get_pet() -> Pet {
    Pet::Dog(Dog::new())
}

fn speak_from_func(speaker: &dyn Voice) {
    // `&dyn Voice` is a wide pointer:
    // 1. Pointer to data
    // 2. Pointer to vtable
    //
    // Also has to do 3 dereferences (if the data is required), but 1. and 2.
    // can be dereferenced simultaneously.
    speaker.speak();
}

fn create_pets() -> Vec<Pet> {
    let mut pets = Vec::with_capacity(2);
    // No extra allocations!
    pets.push(Pet::Dog(Dog::new()));
    pets.push(Pet::Cat(Cat::new()));
    pets
}

fn main() {
    // Normal construction.
    let dog = Dog::new();
    dog.speak();
    let cat = Cat::new();
    cat.speak();

    // Passing to functions.
    speak_from_func(&dog);
    speak_from_func(&cat);

    // Enums.
    let pet = get_pet();
    match pet {
        Pet::Cat(cat) => cat.speak(),
        Pet::Dog(dog) => dog.speak(),
    };
    // pet.speak();

    // Iterating over pets.
    let pets = create_pets();
    for pet in pets.iter() {
        // No deref, only switch.
        pet.speak();
    }
}
